### PR TITLE
V1.2.0

### PR DESCRIPTION
**Pull request by carloalbertobarbano**
Changelog: 
- Add proper support for multiple backends (numpy, torch, tensorflow)
- Add installation flavors in setup.py

### CODE REVIEW

**Comment by carloalbertobarbano** [comment]
@andreped any other pressing matter for this release?

**Comment by andreped** [comment]
> @andreped any other pressing matter for this release?

I don't think so. We have unit tests to verify that everything works across different python, tensorflow, and pytorch-versions, but I have not done a lot of testing locally on projects that depends on torchstain (as it is currently not used that much, as I mostly use TF). But perhaps you could run some tests, just to see that it works properly on your projects?

Otherwise, I think the release is ready. If we find any minor bugs or similar, we could always make a patch release.

**Comment by andreped** [comment]
Which branch am I supposed to look at now? Is it v1.2.0 for this release? I observed that there were two branches for v1.1.0 and v1.2.0, respectively.

**Comment by carloalbertobarbano** [comment]
I created a v1.1.0 just for archiving and quick access, the new version is v1.2.0

**Comment by andreped** [comment]
When you have done the last modifications and are ready to merge with main, I can review the edits to see if there are any improvements to be made (if necessary), if you'd like.

**Comment by carloalbertobarbano** [comment]
Seems like the tf implementation for percentile does not work: https://github.com/EIDOSLAB/torchstain/runs/7728799452?check_suite_focus=true 
However, I am not very familiar with tensorflow

**Comment by carloalbertobarbano** [comment]
I squashed the last 10 commits with fixes for the action yaml

**Comment by andreped** [comment]
> Seems like the tf implementation for percentile does not work

Hmm, what I often observe is that simple methods do not always work identical or have expected/similar behaviour across frameworks. This is exactly why I had to implement my own least square method, as TF did not handle non-singular matrices in the same was as pytorch (see [solveLS](https://github.com/EIDOSLAB/torchstain/blob/v1.2.0/torchstain/tf/normalizers/solveLS.py)).

I can take a look at the percentile function tomorrow, setup some unit tests to see that we get expected behaviour across backends. Please, create an issue for that.

Lastly, I would also say that it makes more sense to have the solveLS inside the utils folder, instead of the normalizer folder where it lies now, but up to you.

**Comment by carloalbertobarbano** [comment]
Okay! 

> Lastly, I would also say that it makes more sense to have the solveLS inside the utils folder

I agree, I missed it

**Comment by carloalbertobarbano** [comment]
I made a RC release [v1.2.0-rc1](https://github.com/EIDOSLAB/torchstain/releases/tag/v1.2.0-rc1). If everything is okay, I will make a final release and update the pypi package.

**Comment by andreped** [comment]
> I made a RC release [v1.2.0-rc1](https://github.com/EIDOSLAB/torchstain/releases/tag/v1.2.0-rc1). If everything is okay, I will make a final release and update the pypi package.

I can take a look now. Will make a PR if I see something crucial. But I would just merge this branch with main, even if we make some minor commits after.

**Review by andreped — CHANGES_REQUESTED** [review]
Comments were written [here](https://github.com/EIDOSLAB/torchstain/pull/18#issuecomment-1209171242). Please, adjust for these comments before merge with main.

**Comment by andreped** [comment]
Just saw that the [TF test failed](https://github.com/EIDOSLAB/torchstain/runs/7744762960?check_suite_focus=true) with Python 3.6, but that is only because newer versions of TF weren't available at the time. You could remove Python 3.6 from the tests, or add specific versions of TF in that test.

In a related project, I did [this](https://github.com/andreped/GradientAccumulator/blob/main/.github/workflows/test.yml#L91), if you wish to keep the Python 3.6 test.

**Comment by carloalbertobarbano** [comment]
I really feel like 164 tests are a bit too many..
I would restrict testing to ubuntu only. We do not make use of system calls and pytorch, tf and numpy should be already tested on all systems.

**Comment by andreped** [comment]
> I really feel like 164 tests are a bit too many..

I believe we should perform these tests before we make a release, but it is not necessary for every single commit.

> I would restrict testing to ubuntu only.

We could do this for each commit. It is also likely not necessary to test multiple different tf/pytorch versions for each commit.

Otherwise, I think the fork is ready to be merged, especially as the last CI ran successfully (see [here](https://github.com/EIDOSLAB/torchstain/actions/runs/2824993391))
